### PR TITLE
feat(pubsub): retry transient failures in ack/modack in timed unary buffer

### PR DESCRIPTION
**Pull request by NivedhaSenthil**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:

- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/google-cloud-ruby/issues) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea.
- [x] Follow the instructions in [CONTRIBUTING](CONTRIBUTING.md). Most importantly, ensure the tests and linter pass by running `bundle exec rake ci` in the gem subdirectory.
- [x] Update code documentation if necessary.

closes: #18235

### CODE REVIEW

**Comment by NivedhaSenthil** [comment]
Tests are failing due to a regression in minitest's latest version(5.16.0). Fixed via https://github.com/googleapis/google-cloud-ruby/pull/18396

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb [hunk 1]
```diff
@@ -92,13 +106,23 @@ def flush!
             with_threadpool do |pool|
               requests[:acknowledge].each do |ack_req|
                 add_future pool do
-                  @subscriber.service.acknowledge ack_req.subscription, *ack_req.ack_ids
+                  begin
+                    @subscriber.service.acknowledge ack_req.subscription, *ack_req.ack_ids
+                  rescue *RETRIABLE_ERRORS => e
+                    (handle_failure e, ack_req.ack_ids) if @subscriber.exactly_once_delivery_enabled
```
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb:112]: I don't think the parens are needed.

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb [hunk 2]
```diff
@@ -129,6 +154,60 @@ def stopped?
 
           private
 
+          def handle_failure error, ack_ids, ack_deadline_seconds = nil
+            ack_ids = (parse_error error) || ack_ids
```
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb:158]: I think it's more ruby-like to say `parse_error(error) || ack_ids`.

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb [hunk 3]
```diff
@@ -129,6 +154,60 @@ def stopped?
...
+            return temporary_failures.keys.map(&:to_s) unless temporary_failures.empty?
+          end
+
+          def handle_permanent_failures error
+            # TODO: Add log or pass on result(AcknowledgeResult) to callback
+          end
+
+          def perform_retry_async ack_ids, ack_deadline_seconds = nil
+            return unless retry_thread_pool.running?
+            Concurrent::Promises.future_on(
```
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb:178]: What happens if the thread pool is shut down between the above two lines?

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb [hunk 4]
```diff
@@ -129,6 +154,60 @@ def stopped?
...
+            return if metadata.nil?
+            permanent_failures, temporary_failures = metadata.partition do |_, v|
+              v.include? PERMANENT_FAILURE
+            end.map(&:to_h)
+            handle_permanent_failures permanent_failures
+            return temporary_failures.keys.map(&:to_s) unless temporary_failures.empty?
+          end
+
+          def handle_permanent_failures error
+            # TODO: Add log or pass on result(AcknowledgeResult) to callback
```
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb:173]: Include a link to an open issue for this.

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb [hunk 5]
```diff
@@ -129,6 +154,60 @@ def stopped?
...
+          end
+
+          def parse_error error
+            metadata = error.error_metadata
+            return if metadata.nil?
+            permanent_failures, temporary_failures = metadata.partition do |_, v|
+              v.include? PERMANENT_FAILURE
+            end.map(&:to_h)
+            handle_permanent_failures permanent_failures
+            return temporary_failures.keys.map(&:to_s) unless temporary_failures.empty?
```
Comment (dazuma) [inline google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb:169]: `return` is redundant here. (The `unless` guard already evaluates to `nil` if it fails.)

File: google-cloud-pubsub/test/google/cloud/pubsub/subscriber/timed_unary_buffer_test.rb [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+    rec_message_hash("rec_message1-msg-goes-here", 1111) }
+  let(:rec_msg2_grpc) { Google::Cloud::PubSub::V1::ReceivedMessage.new \
+    rec_message_hash("rec_message2-msg-goes-here", 1112) }
+  let(:rec_msg3_grpc) { Google::Cloud::PubSub::V1::ReceivedMessage.new \
+    rec_message_hash("rec_message3-msg-goes-here", 1113) }
+
+
+  it "should call handle error for ack on retriable error" do
+    pull_res1 = Google::Cloud::PubSub::V1::StreamingPullResponse.new received_messages: [rec_msg1_grpc],
+                                                                    subscription_properties: {
```
Comment (dazuma) [inline google-cloud-pubsub/test/google/cloud/pubsub/subscriber/timed_unary_buffer_test.rb:34]: Please line up the indentation (also on a number of similar calls below).

File: google-cloud-pubsub/test/google/cloud/pubsub/subscriber/timed_unary_buffer_test.rb [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    end
+  
+    subscription.service.mocked_subscriber = stub
+    subscriber = subscription.listen streams: 1 do |msg|
+      msg.modify_ack_deadline! 120
+      called = true
+    end
+
+    subscriber.on_error do |error|
+        errors << error
```
Comment (dazuma) [inline google-cloud-pubsub/test/google/cloud/pubsub/subscriber/timed_unary_buffer_test.rb:131]: Indent 2 spaces. (This also appears several more times below, sometimes on blocks, and sometimes on embedded methods.)

**Review by dazuma — CHANGES_REQUESTED** [review]

File: google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb (hunk 3 above)
Comment (NivedhaSenthil) [reply google-cloud-pubsub/lib/google/cloud/pubsub/subscriber/timed_unary_buffer.rb:179]: It throws `#<Concurrent::RejectedExecutionError: Concurrent::RejectedExecutionError>` which gets propagated via the error callback registered on subscriber. But this should be a rare scenario as retry_thread_pool would be shutdown only on a stop call on the subscriber and all stream shutdown are waited before shutting the buffer.

**Pull request merged**